### ISSUE TITLE
Unbounded layout of `group`ed Line fails

### ISSUE THREAD

**Issue by sjakobi** [quchen/prettyprinter#91]
```haskell
> layoutUnbounded = layoutPretty defaultLayoutOptions { layoutPageWidth = Unbounded }
> layoutUnbounded $ group Line
SFail
```

**Comment by sjakobi** [comment]
The possibility of `Fail` is introduced by `group` via `changesUponFlattening`:

```haskell
> group Line
Union Fail Line
```

https://github.com/quchen/prettyprinter/blob/4d51362d7ea8bca0fe9e7ffec2ccd8fcd0065c0d/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L552-L555

The `Unbounded` layout causes the  maximum width to set to `Nothing`:

https://github.com/quchen/prettyprinter/blob/4d51362d7ea8bca0fe9e7ffec2ccd8fcd0065c0d/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L1789-L1803

`selectNicer` then selects `Fail` because the fitting predicates always choose the first option when the maximum width is `Nothing`:

https://github.com/quchen/prettyprinter/blob/4d51362d7ea8bca0fe9e7ffec2ccd8fcd0065c0d/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L1644-L1647

https://github.com/quchen/prettyprinter/blob/4d51362d7ea8bca0fe9e7ffec2ccd8fcd0065c0d/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L1710-L1713

**Comment by sjakobi** [comment]
I assume there are good reasons for `group` to introduce the possibility of `Fail`. So I think a reasonable fix would be to let `selectNicer` to choose `y` when `x` is `Fail`.

**Comment by Gabriel439** [comment]
@sjakobi: My understanding is that this behavior is intentional for `group Line` to fail because it needs to contract the string to a newline-free version.  See the documentation for `group`:

> (group x) tries laying out x into a single line by removing the contained line breaks;

In other words, you need to protect all occurrences of `Line` (or `hardline`, which is equivalent) with `flatAlt` to provide a newline-free alternative that `group` can fall back to.  The alternative cannot itself contain obligatory newlines.

`dhall` had a similar issue in the past due to misusing `group` in this way:

https://github.com/dhall-lang/dhall-haskell/pull/272

**Comment by sjakobi** [comment]
Interesting!

Note that bounded layouts don't fail though:

```
> layoutPretty defaultLayoutOptions $ group Line
SLine 0 SEmpty
> layoutSmart defaultLayoutOptions $ group Line
SLine 0 SEmpty
```

And I don't really see why the layout _ought_ to fail, when the failure of the first alternative can be detected and the second alternative can be returned instead.

**Comment by sjakobi** [comment]
Also, the documentation of `hardline` shows an example where the line break is not protected via `flatAlt`:

https://github.com/quchen/prettyprinter/blob/4d51362d7ea8bca0fe9e7ffec2ccd8fcd0065c0d/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L504-L517

**Comment by Gabriel439** [comment]
@sjakobi: I think it is because `prettyprinter` is interpreting "unbounded" layout as "render on one line at all costs because there is no right margin", and then fails if there is an obligatory newline in the document

**Comment by sjakobi** [comment]
Hmm. I still find this interpretation very hard to believe, not least because I don't see any documentation supporting it.

For example this documentation for `Unbounded` seems to indicate that users _may_ introduce hard line breaks, while e.g. `softline` will always be layouted not to introduce a line break.

https://github.com/quchen/prettyprinter/blob/4d51362d7ea8bca0fe9e7ffec2ccd8fcd0065c0d/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L1610-L1611

Seems like the documentation needs more clarification though!

**Comment by Gabriel439** [comment]
@sjakobi: I agree that the documentation and/or the `Unbounded` name are perhaps misleading, but it's actually a reasonable behavior if you ask me.  Otherwise there would be no way for the user to specify that they want to force a single-line representation.

**Comment by sjakobi** [comment]
Here's another bit that IMO indicates that this behaviour is _not_ intended:

https://github.com/quchen/prettyprinter/blob/4d51362d7ea8bca0fe9e7ffec2ccd8fcd0065c0d/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L84-L88

---

Even if it were intended, I'd like to argue that the behaviour should be changed so

    render* . layout{Pretty,Simple} (LayoutOptions Unbounded) . group

is a total function as long as the input `Doc` doesn't contain `Fail`:

In some situations like https://github.com/dhall-lang/dhall-haskell/pull/1508 you want a `Doc` that is unconditionally rendered with a line break.

As long as layouting and rendering that document can result in an `error`, it seems that it would be the responsibility of the author of that `Doc` to warn users of this possibility. That would IMHO be a disincentive to rely on `prettyprinter` at all.

**Comment by sjakobi** [comment]
I'd like to fix this soon, probably by merging #98.

I wonder whether I should try to accommodate users, who, like @Gabriel439, rely on the error to detect multi-line documents. For example we could add a predicate like

    isSingleLine :: SimpleDocStream ann -> Bool

Thoughts?

**Comment by sjakobi** [comment]
Note that #116 also fixes the original testcase, but already on the `Doc`-level:

```haskell
> diag $ group hardline 
Line
```

`Fail` is avoided entirely.

However `dhall`'s testsuite will still fail when I use `Unbounded` in its layout options:

```patch
--- a/dhall/src/Dhall/Pretty/Internal.hs
+++ b/dhall/src/Dhall/Pretty/Internal.hs
@@ -1349,7 +1349,7 @@ layout = Pretty.removeTrailingWhitespace . Pretty.layoutSmart layoutOpts
 layoutOpts :: Pretty.LayoutOptions
 layoutOpts =
     Pretty.defaultLayoutOptions
-        { Pretty.layoutPageWidth = Pretty.AvailablePerLine 80 1.0 }
+        { Pretty.layoutPageWidth = Pretty.Unbounded }
 
 {- $setup
 >>> import Test.QuickCheck (Fun(..))
```
